### DISCUSSION TITLE
How to require email verification only when user has filled his email address?

### DISCUSSION THREAD

**Discussion by renatofrota** [filamentphp/filament#14044] [Help] [answered]
### Package

Panel builder

### Package Version

v3.2.102

### How can we help you?

I have customized the login process to accept login+password. Now, I would like to require email verification only when user has updated their profile and added email address.

Adding emailVerification() to $panel on AdminPanelProvider triggers the verification even when user has no email address at all.

Also, user can't access profile page when email verification is required - so if inserting a wrong email address, he can't update it.

Optimaly:

- emailVerification() could have an option to ignore verification process when user has no email address at all;
- profile page could be accessed before verification is completed so user could update email address if wrong;

**Comment by e2d2-dev** [comment]
Hi,

->emailVerification() accepts a Closure, so fn() => Auth::user()->email should do the job (didn't tested it).


ciao ed

**Reply by renatofrota** [reply]
thanks but the closure is for the first parameter. isRequired is the second one.

and for some reason I can't use Auth:: or auth() within the panel() function.

**Reply by e2d2-dev** [reply]
use 

```
->emailVerification(
    isRequired: fn() => yourFunction
)
```

**Reply by renatofrota** [reply]
It doesn't work. Although the closure returns a bool, the error message states it should be a boolean (directly, not through closure).

However, I've found a solution (I will reply separately). Thanks

**Comment by renatofrota** [comment, marked as answer]
Resolved adding to \App\Models\Users:

```php
public function hasVerifiedEmail()
{
    return is_null($this->email) or ! is_null($this->email_verified_at) or request()->routeIs('filament.*.auth.profile');
}
```

- user has no email: is considered to be validated;
- user provides email: validation is required;
- profile page is accessible before verification so email may be updated if necessary;

**Reply by renatofrota** [reply]
Additional information:

In order to remove `email_verified_at` timestamp when user modify/clear email, I've created a custom EditProfile class `->profile(CustomProfile::class)` as follows:

```php
<?php

namespace App\Filament\Pages\Auth;

use Filament\Forms\Form;
use Filament\Pages\Auth\EditProfile as FilamentEditProfile;
use Illuminate\Database\Eloquent\Model;

class CustomProfile extends FilamentEditProfile
{
    public function form(Form $form): Form
    {
        return $form
            ->schema([
                $this->getNameFormComponent(),
                $this->getEmailFormComponent()
                    ->required(false),
            ])
            ->statePath('data');
    }

    protected function handleRecordUpdate(Model $record, array $data): Model
    {
        if ($data['email'] === null or $data['email'] !== $record->email) {
            $data['email_verified_at'] = null;
        }

        $record->update($data);

        return $record;
    }

}
```

my last doubt now is how to send verification email automatically when user changes their email (with no need to click "resend" link manually).

**Reply by e2d2-dev** [reply]
Just add a listener to the User Model with $user->notify(Mailable)
